File: 082_Binary Tree Right Side View.cpp
// Binary Tree Right Side View
// Return values seen from the right side, level by level.
#include <bits/stdc++.h>
using namespace std;
struct TreeNode{int val; TreeNode* left; TreeNode* right; TreeNode(int x): val(x), left(nullptr), right(nullptr) {}};

vector<int> rightSideView(TreeNode* root){
    vector<int> res;
    if (!root) return res;
    queue<TreeNode*> q; q.push(root);
    while(!q.empty()){
        int n=q.size();
        for(int i=0;i<n;++i){
            TreeNode* t=q.front(); q.pop();
            if(i==n-1) res.push_back(t->val);
            if(t->left) q.push(t->left);
            if(t->right) q.push(t->right);
        }
    }
    return res;
}

int main(){
    TreeNode* root = new TreeNode(1);
    root->right = new TreeNode(3);
    root->left = new TreeNode(2);
    root->left->right = new TreeNode(5);
    root->right->right = new TreeNode(4);
    auto v = rightSideView(root);
    cout<<"Right side view: "; for(int x:v) cout<<x<<" "; cout<<"\n";
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(n)

File: 082_Binary Tree Right Side View.py
"""
Binary Tree Right Side View
Return node values visible from the right side, level by level.
"""
from typing import Optional, List

class TreeNode:
    def __init__(self,val=0,left=None,right=None):
        self.val=val; self.left=left; self.right=right

def right_side_view(root: Optional[TreeNode]) -> List[int]:
    if not root: return []
    from collections import deque
    q = deque([root])
    res = []
    while q:
        n = len(q)
        for i in range(n):
            node = q.popleft()
            if i==n-1: res.append(node.val)
            if node.left: q.append(node.left)
            if node.right: q.append(node.right)
    return res

if __name__ == '__main__':
    root = TreeNode(1)
    root.right = TreeNode(3)
    root.left = TreeNode(2)
    root.left.right = TreeNode(5)
    root.right.right = TreeNode(4)
    print('Right side view:', right_side_view(root))

# Time Complexity: O(n)
# Space Complexity: O(n)
